### PR TITLE
Add xformers attention to VAE

### PR DESCRIPTION
**Pull request by kig**
This PR adds xformers memory-efficient attention to the VAE to make the attention parts of the VAE require less memory.

This can help with issues like #1434 to some extent, but another part of the problem are the convolution layers. Running the VAE in channels_last memory format helps, but you're still dealing with RAM use in the tens of GB . At least it's not hundreds of GB. To run the VAE on low VRAM, switch the up_blocks / down_blocks to run on the CPU.

### CODE REVIEW

**Comment by blefaudeux** [comment]
thoughts on https://github.com/huggingface/diffusers/pull/1493 @kig ? It's basically meant to help PRs like this one, checking just in case that it would also make sense to somebody else who spent time in these layers

File: src/diffusers/models/attention.py [hunk 1]
```diff
@@ -290,6 +290,33 @@ def __init__(
         self.rescale_output_factor = rescale_output_factor
         self.proj_attn = nn.Linear(channels, channels, 1)
 
+        self._use_memory_efficient_attention_xformers = False
+
+    def _set_use_memory_efficient_attention_xformers(self, use_memory_efficient_attention_xformers: bool):
```
Comment (patil-suraj) [inline src/diffusers/models/attention.py:295]: Let's make this public. And thanks to #1493, it's the only place where we need to add this method!

```suggestion
    def set_use_memory_efficient_attention_xformers(self, use_memory_efficient_attention_xformers: bool):
```

File: src/diffusers/models/attention.py [hunk 2]
```diff
@@ -324,21 +351,26 @@ def forward(self, hidden_states):
...
+                    dtype=query_proj.dtype,
+                    device=query_proj.device,
+                ),
+                query_proj,
+                key_proj.transpose(-1, -2),
+                beta=0,
+                alpha=scale,
+            )
+            attention_probs = torch.softmax(attention_scores.float(), dim=-1).type(attention_scores.dtype)
+            hidden_states = torch.bmm(attention_probs, value_proj)
```
Comment (patil-suraj) [inline src/diffusers/models/attention.py:373]: Looks good!

File: src/diffusers/models/unet_2d_blocks.py [hunk 3]
```diff
@@ -305,6 +305,10 @@ def __init__(
         self.attentions = nn.ModuleList(attentions)
         self.resnets = nn.ModuleList(resnets)
 
+    def set_use_memory_efficient_attention_xformers(self, use_memory_efficient_attention_xformers: bool):
+        for attn in self.attentions:
+            attn._set_use_memory_efficient_attention_xformers(use_memory_efficient_attention_xformers)
+
```
Comment (patil-suraj) [inline src/diffusers/models/unet_2d_blocks.py:311]: This can be removed!

File: src/diffusers/models/vae.py [hunk 4]
```diff
@@ -125,6 +125,12 @@ def __init__(
         conv_out_channels = 2 * out_channels if double_z else out_channels
         self.conv_out = nn.Conv2d(block_out_channels[-1], conv_out_channels, 3, padding=1)
 
+    def set_use_memory_efficient_attention_xformers(self, use_memory_efficient_attention_xformers: bool):
+        self.mid_block.set_use_memory_efficient_attention_xformers(use_memory_efficient_attention_xformers)
+        for block in self.down_blocks:
+            if hasattr(block, "attentions") and block.attentions is not None:
+                block.set_use_memory_efficient_attention_xformers(use_memory_efficient_attention_xformers)
+
```
Comment (patil-suraj) [inline src/diffusers/models/vae.py:133]: This can be removed!

File: src/diffusers/models/vae.py [hunk 5]
```diff
@@ -205,6 +211,12 @@ def __init__(
         self.conv_act = nn.SiLU()
         self.conv_out = nn.Conv2d(block_out_channels[0], out_channels, 3, padding=1)
 
+    def set_use_memory_efficient_attention_xformers(self, use_memory_efficient_attention_xformers: bool):
+        self.mid_block.set_use_memory_efficient_attention_xformers(use_memory_efficient_attention_xformers)
+        for block in self.up_blocks:
+            if hasattr(block, "attentions") and block.attentions is not None:
+                block.set_use_memory_efficient_attention_xformers(use_memory_efficient_attention_xformers)
```
Comment (patil-suraj) [inline src/diffusers/models/vae.py:218]: This can be removed!

File: src/diffusers/models/vae.py [hunk 6]
```diff
@@ -567,6 +579,10 @@ def __init__(
         self.post_quant_conv = torch.nn.Conv2d(latent_channels, latent_channels, 1)
         self.use_slicing = False
 
+    def set_use_memory_efficient_attention_xformers(self, use_memory_efficient_attention_xformers: bool):
+        self.decoder.set_use_memory_efficient_attention_xformers(use_memory_efficient_attention_xformers)
+        self.encoder.set_use_memory_efficient_attention_xformers(use_memory_efficient_attention_xformers)
+
```
Comment (patil-suraj) [inline src/diffusers/models/vae.py:585]: This can be removed!

**Comment by patil-suraj** [comment]
Hey @kig , thanks a lot for the PR! The PR looks good, we just need to add the `set_use_memory_efficient_attention_xformers` only in `AttentionBlock` now thanks to #1493. So we can remove this method from other places. Should be good to merge after that :)

File: src/diffusers/models/attention.py (hunk 1 above)
Comment (kig) [reply src/diffusers/models/attention.py]: Thanks, done!

**Comment by kig** [comment]
Hi @patil-suraj , thanks for the help! I applied the changes and it's a lot simpler now :)

File: src/diffusers/models/attention.py [hunk 7]
```diff
@@ -286,6 +286,33 @@ def __init__(
         self.rescale_output_factor = rescale_output_factor
         self.proj_attn = nn.Linear(channels, channels, 1)
 
+        self._use_memory_efficient_attention_xformers = False
+
+    def set_use_memory_efficient_attention_xformers(self, use_memory_efficient_attention_xformers: bool):
+        if not is_xformers_available():
+            print("Here is how to install it")
```
Comment (patil-suraj) [inline src/diffusers/models/attention.py:293]:
```suggestion
```

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for addressing the comments!

**Comment by patil-suraj** [comment]
Tried it out, works really well, merging!

**Pull request merged**